subsys/mgmt/mcumgr/smp.c

```diff
@@ -117,6 +117,9 @@ zephyr_smp_split_frag(struct net_buf **nb, void *arg, uint16_t mtu)
 		frag = src;
 	} else {
 		frag = zephyr_smp_alloc_rsp(src, arg);
+		if (!frag) {
+			return NULL;
+		}
```

Comment:
This prevents NULL dereference, but caller of `zephyr_smp_split_frag()` will then leak buffer:

https://github.com/zephyrproject-rtos/zephyr/blob/d00b37c8dd19d8e519baf20ed3b9ed1bdea28620/subsys/mgmt/mcumgr/smp.c#L185-L188

Reply:
The point of this PR was not to fix everything within the smp implementation. For my use case i was getting bus faults which i want to fix. Im not maintainer of mcumgr. I could create an issue for that particular problem. Would that be okay?

Reply:
I don't see much value of fixing a bus fault, when instead of **each** fixed bus fault you will end up with one buffer leaked. You are just changing one type of error (bus fault) to another one (buffer leak). It is questionable which type of error is worse, but IMO buffer leaks take more time to spot and fix than bus fault (which have a dedicated handler and nice error log on UART).

I think that those two are too closely related, so they should be fixed within single PR.

Reply:
Bus fault prevents operation altogether so yes, it is easier to spot. With leak you can live as long as there is enough memory.
Anyway, let's try to understand the leak. @mniestroj , can you tell which buffer will leak exactly?

Reply:
btw - where is unref done on netbuf? The function documentation refers to one being done (https://github.com/zephyrproject-rtos/zephyr/blob/d00b37c8dd19d8e519baf20ed3b9ed1bdea28620/subsys/mgmt/mcumgr/smp.c#L89).
It seems there are also other things in this file that could be cleaned up... anyway this is not the issue for this PR.

Reply:
I don't think im changing bus fault into buffer leak as smp_bt works just fine. From what i understand looking into the code very briefly. The only thing happening there would be not sending the response.

Here the response callback is initialized.
```
	int rc;

	streamer = (struct smp_streamer) {
		.mgmt_stmr = {
			.cfg = &zephyr_smp_cbor_cfg,
			.reader = &reader.r,
			.writer = &writer.enc,
			.cb_arg = zst,
		},
		.tx_rsp_cb = zephyr_smp_tx_rsp,
	};
```

Here used
```
        /* Send the response. */
        rc = streamer->tx_rsp_cb(streamer, rsp, streamer->mgmt_stmr.cb_arg);
        rsp = NULL;
        if (rc != 0) {
            break;
        }

        /* Trim processed request to free up space for subsequent responses. */
        mgmt_streamer_trim_front(&streamer->mgmt_stmr, req,
                                 smp_align4(req_hdr.nh_len));

        cmd_done_arg.err = MGMT_ERR_EOK;
        mgmt_evt(MGMT_EVT_OP_CMD_DONE, req_hdr.nh_group, req_hdr.nh_id,
                 &cmd_done_arg);
    }

    if (rc != 0 && valid_hdr) {
        smp_on_err(streamer, &req_hdr, req, rsp, rc);

        if (handler_found) {
            cmd_done_arg.err = rc;
            mgmt_evt(MGMT_EVT_OP_CMD_DONE, req_hdr.nh_group, req_hdr.nh_id,
                     &cmd_done_arg);
        }

        return rc;
    }
```

Seems like smp can handle that well.

Reply:
> With leak you can live as long as there is enough memory.

There are only `CONFIG_MCUMGR_BUF_COUNT` (4 by default) buffers.

In `zephyr_smp_split_frag()` you do:
```
		frag = zephyr_smp_alloc_rsp(src, arg);
		if (!frag) {
			return NULL;
		}
```
and the caller (`zephyr_smp_tx_rsp()`) does following:
```
		frag = zephyr_smp_split_frag(&nb, zst, mtu);
		if (frag == NULL) {
			return MGMT_ERR_ENOMEM;
		}
```
If `zephyr_smp_alloc_rsp()` returned `NULL`, then `zephyr_smp_split_frag()` will also return NULL. The problem is, that then `MGMT_ERR_ENOMEM` is returned immediately and `nb` is not processed / consumed to the end. So we need to free `nb` in case `zephyr_smp_alloc_rsp()` and then `zephyr_smp_split_frag()` return NULL.

Normally `nb` is consumed by `rc = zst->zst_output(zst, frag);`. I think we should call `zephyr_smp_free_buf()` on `nb` in case of split error:
```
		frag = zephyr_smp_split_frag(&nb, zst, mtu);
		if (frag == NULL) {
			zephyr_smp_free_buf(nb);
			return MGMT_ERR_ENOMEM;
		}
```

Reply:
thanks

Reply:
@mniestroj Updated with your suggestions, please have a look.